### PR TITLE
Fix movable-to-world collision ignoring initial state

### PR DESCRIPTION
**Pull request by williamshen-nz**
## Summary

- Objects resting on surfaces may be in collision at their initial pose (`pose_ts=0`) due to perception noise (e.g., detected pose slightly penetrates the surface). The movable-to-movable check already handles this by filtering from `obj_to_first_pose_ts` onward, but `movable_to_world` did not — it penalized all particles uniformly at all timesteps including the initial one.
- Batch all activated objects into a single `collision_fn` call using `torch.stack` + `einops.rearrange`, then apply a cached mask to zero out timesteps before each object's first placement, consistent with the movable-to-movable approach.

## Changes

- Replaced the old concatenate-all-spheres approach in `collision_costs()` with a vectorized batch: `torch.stack` all activated objects' spheres, reshape via `einops.rearrange` to `(objs * b, t, n, 4)` for a single `collision_fn` call, reshape back, then mask and sum over objects.
- Mask tensor `(num_objs, 1, t)` is cached after first call since it's fixed per plan skeleton.
- Sorted activated objects list cached in `__init__` for deterministic `torch.stack` ordering.
- Added test environment (`blocks_activation_dist_test.yml`) with blocks slightly penetrating the floor (simulating perception noise).
- Added test script (`test_movable_to_world_fix.py`) that monkey-patches the old code to verify the fix.

## Test plan

- [x] `test_movable_to_world_fix.py` — Old code: 0/512 satisfying (FAIL). New code: 375/512 satisfying (PASS)
- [x] `cutamp-demo --env tetris_1` (single object) — runs successfully
- [x] `cutamp-demo --env tetris_3` (3 objects, 512 particles, 500 steps) — 7/512 satisfying
- [x] `cutamp-demo --env tetris_5` (5 objects, 512 particles, 500 steps) — 1/512 satisfying
- [x] `cutamp-demo --env blocks` (4 objects, 512 particles, 500 steps) — 235/512 satisfying
- [x] `cutamp-demo --env unpack` (3 objects, 512 particles, 500 steps) — runs without errors
- [ ] Test on real robot
- [ ] Make sure optimization times did not decrease too substantially

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by williamshen-nz** [comment]
Updated the implementation to address the vectorization concern:

**Before (this PR, prior commit):** Per-object Python loop calling `collision_fn` individually + `sum()` to aggregate  
**After (latest commit):** Single batched `collision_fn` call using `torch.stack` + `reshape`, then mask tensor + `sum(0)`

The small for-loop that remains is only for building the mask tensor (setting zeros for pre-placement timesteps), which is O(num_objects) on CPU and negligible compared to the GPU collision call.

All tests pass:
- `tetris_1` (64 particles, 50 steps): 9/64 satisfying
- `tetris_3` (64 particles, 100 steps): runs without errors
- `test_movable_to_world_fix`: old code FAIL, new code PASS

File: cutamp/cost_function.py [hunk 1]
```diff
@@ -437,9 +440,24 @@ def collision_costs(self, rollout: Rollout, obj_to_spheres: Dict[str, Float[torc
...
+        coll = self.world.collision_fn(rearrange(stacked, "objs b t n d -> (objs b) t n d"))
+        coll = rearrange(coll, "(objs b) t -> objs b t", objs=len(self._activated_objs))
+
+        # Mask is fixed per skeleton — cache after first call
+        if self._movable_world_mask is None:
+            num_objs, t = coll.shape[0], coll.shape[2]
+            mask = torch.ones(num_objs, 1, t, device=coll.device)
+            for i, obj in enumerate(self._activated_objs):
+                first_ts = self.obj_to_first_pose_ts[obj]
+                if first_ts > 0:
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/cost_function.py]: Acknowledged — Pick-only skeletons aren't supported yet but we'll handle this when they are. For now a KeyError is preferable to silently defaulting to 0 which could mask bugs.

File: cutamp/cost_function.py [hunk 2]
```diff
@@ -437,9 +440,24 @@ def collision_costs(self, rollout: Rollout, obj_to_spheres: Dict[str, Float[torc
...
+        stacked = torch.stack([obj_to_spheres[obj] for obj in self._activated_objs])
+        coll = self.world.collision_fn(rearrange(stacked, "objs b t n d -> (objs b) t n d"))
+        coll = rearrange(coll, "(objs b) t -> objs b t", objs=len(self._activated_objs))
+
+        # Mask is fixed per skeleton — cache after first call
+        if self._movable_world_mask is None:
+            num_objs, t = coll.shape[0], coll.shape[2]
+            mask = torch.ones(num_objs, 1, t, device=coll.device)
+            for i, obj in enumerate(self._activated_objs):
+                first_ts = self.obj_to_first_pose_ts[obj]
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/cost_function.py]: CostFunction is instantiated per skeleton and t/device/dtype are fixed throughout optimization. Not addressing.

File: cutamp/envs/assets/blocks_activation_dist_test.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Test environment for movable-to-world initial collision edge case.
+# Blocks are positioned so their collision spheres fall within a non-zero
+# world_activation_distance of the floor surface, triggering false initial
+# collision costs that the fix in _movable_to_world_cost should mask out.
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/envs/assets/blocks_activation_dist_test.yml]: Fixed in 9c1271b.

File: cutamp/scripts/test_movable_to_world_fix.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
+"""
+Test script to verify the movable-to-world initial collision fix.
+
+Creates an environment where blocks rest on a surface within the
+world_activation_distance, causing false initial collision costs.
+With the old code (no masking), movable_to_world collision penalizes
+all particles uniformly at pose_ts=0, making it much harder to find
+a feasible plan. With the fix, initial timesteps are masked per-object.
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/scripts/test_movable_to_world_fix.py]: Fixed in 9c1271b.

File: cutamp/scripts/test_movable_to_world_fix.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    cost_reducer = CostReducer(default_constraint_to_mult.copy())
+    constraint_checker = ConstraintChecker(default_constraint_to_tol.copy())
+
+    # Monkey-patch to toggle the fix on/off
+    if not use_fix:
+        import torch
+        from cutamp.cost_function import CostFunction
+
+        _original_collision_costs = CostFunction.collision_costs
+
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/scripts/test_movable_to_world_fix.py]: This is a standalone test script, not a library — if it crashes the process exits anyway. Not addressing.

File: cutamp/cost_function.py [hunk 6]
```diff
@@ -437,9 +440,24 @@ def collision_costs(self, rollout: Rollout, obj_to_spheres: Dict[str, Float[torc
...
 
-        # Collision between movables and world
-        activated_movable_spheres = torch.cat([obj_to_spheres[obj] for obj in self.activated_obj], dim=2)
-        coll_values["movable_to_world"] = self.world.collision_fn(activated_movable_spheres)
+        # Collision between movables and world — batch all activated objects in one collision_fn call,
+        # then mask out timesteps before each object's first placement (objects may initially be in
+        # collision with surfaces they rest on, e.g. due to perception noise)
+        stacked = torch.stack([obj_to_spheres[obj] for obj in self._activated_objs])
+        coll = self.world.collision_fn(rearrange(stacked, "objs b t n d -> (objs b) t n d"))
+        coll = rearrange(coll, "(objs b) t -> objs b t", objs=len(self._activated_objs))
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/cost_function.py:450]: All movable objects get the same number of collision spheres — `sample_greedy_surface_spheres` is called with a fixed `n_spheres=coll_n_spheres` (default 50) for every object. `n` is always uniform across movables so `torch.stack` is safe.

File: cutamp/scripts/test_movable_to_world_fix.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            robot_spheres = rollout["robot_spheres"]
+            coll_values = {"robot_to_world": self.world.collision_fn(robot_spheres)}
+
+            # OLD CODE: concatenate all activated objects, no masking
+            activated_movable_spheres = torch.cat(
+                [obj_to_spheres[obj] for obj in self.activated_obj], dim=2
+            )
+            coll_values["movable_to_world"] = self.world.collision_fn(activated_movable_spheres)
+
+            # Rest is identical — call original and overwrite the collision values we changed
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/scripts/test_movable_to_world_fix.py]: Replaced monkey-patching with the mask_initial_movable_world_collision config flag in 0039888.

File: cutamp/scripts/test_movable_to_world_fix.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        num_particles=num_particles,
+        robot="fr3_robotiq",
+        num_opt_steps=500,
+        max_loop_dur=max_duration,
+        enable_visualizer=False,
+        rr_spawn=False,
+        enable_experiment_logging=False,
+        # Default activation distance — initial collision is from blocks penetrating the floor
+        # (simulating perception noise), not from the activation distance buffer
+        world_activation_distance=0.0,
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/scripts/test_movable_to_world_fix.py]: Done in 0039888 — test now uses mask_initial_movable_world_collision=True/False directly.

File: cutamp/cost_function.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply cutamp/cost_function.py:450]: Already addressed — see earlier reply.

**Comment by williamshen-nz** [comment]
**Performance note on the `torch.stack` + `rearrange` approach vs the old `torch.cat`:**

The old code called `collision_fn` with `(b, t, n*num_objs, 4)` (all spheres concatenated). The new code calls it with `(num_objs*b, t, n, 4)` (objects flattened into batch dim). Both result in the same total number of sphere-OBB checks in the CUDA kernel (`SdfSphereOBB`). The kernel sums across the sphere dim (`dim=-1`) via `weight_distance` with `sum_distance=True`, so `sum(collision_fn(each_obj))` == `collision_fn(cat(all_objs))` since sum is associative. The extra `.sum(dim=0)` on `(num_objs, b, t)` is trivially cheap. No meaningful performance difference — verified by reading through cuRobo's `PrimitiveCollisionCost.discrete_fn` and `WorldPrimitiveCollision.get_sphere_distance`.

File: tests/test_cost_function.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        enable_visualizer=False,
+        rr_spawn=False,
+        enable_experiment_logging=False,
+        world_activation_distance=0.0,
+        mask_initial_movable_world_collision=mask,
+    )
+    cost_reducer = CostReducer(default_constraint_to_mult.copy())
+    constraint_checker = ConstraintChecker(default_constraint_to_tol.copy())
+    _, num_satisfying, _ = run_cutamp(env, config, cost_reducer, constraint_checker)
+    return num_satisfying
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply tests/test_cost_function.py:35]: The result is extremely binary — mask=False always gets 0/512, mask=True consistently gets 300-400/512. We've run it 5+ times without variation. Seeding wouldn't help much anyway since CUDA ops aren't fully deterministic.

**Review by NishanthJKumar — APPROVED** [review]
lgtm!
